failure with a pinned opam package

I try to use opam monorepo 0.3.3 with the following pins:
`gmp.6.2.1-2  (uninstalled)  git  git+file:///usr/home/hannes/devel/mirage/ocaml-gmp#HEAD`

And receive a failure during `opam monorepo lock`:
```
gmake[2]: Leaving directory '/usr/home/hannes/devel/mirage/mirage-skeleton/applications/static_website_tls'
 ↳ generate lockfile for monorepo dependencies
==> Using 1 locally scanned package as the target.
==> Found 130 opam dependencies for the target package.
==> Querying opam database for their metadata and Dune compatibility.
==> Calculating exact pins for each of them.
opam-monorepo: [ERROR] No HEAD ref for file:///usr/home/hannes/devel/mirage/ocaml-gmp
```

It would be pretty nice to support pins to git repositories the same as opam does, or are there any blockers?

Ah I guess my problem was that I tried running `make lock` in the toplevel:

```
/bin/sh: line 1: cd: tutorial/noop/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: tutorial/noop-functor/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: tutorial/hello/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: tutorial/hello-key/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: tutorial/lwt/echo_server/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: tutorial/lwt/heads1/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: tutorial/lwt/heads2/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: tutorial/lwt/timeout1/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: tutorial/lwt/timeout2/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: tutorial/lwt/echo_server/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: tutorial/app_info/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: device-usage/clock/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: device-usage/conduit_server/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: device-usage/console/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: device-usage/http-fetch/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: device-usage/kv_ro/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: device-usage/network/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: device-usage/ping6/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: device-usage/prng/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: applications/docteur/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: applications/dhcp/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: applications/git/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: applications/dns/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: applications/crypto/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: applications/static_website_tls/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
/bin/sh: line 1: cd: device-usage/block/mirage: No such file or directory
ls: cannot access '*.opam': No such file or directory
[opam-overlays] no changes from git+https://github.com/dune-universe/opam-overlays.git
[NOTE] Repository opam-overlays has been added to the selections of switch /home/me/mirage-skeleton only.
       Run `opam repository add opam-overlays --all-switches|--set-default' to use it in all existing switches, or in newly created switches, respectively.

[mirage-opam-overlays] no changes from git+https://github.com/dune-universe/mirage-opam-overlays.git
[NOTE] Repository mirage-opam-overlays has been added to the selections of switch /home/me/mirage-skeleton only.
       Run `opam repository add mirage-opam-overlays --all-switches|--set-default' to use it in all existing switches, or in newly created switches, respectively.

env OPAMVAR_monorepo="opam-monorepo" opam monorepo lock --recurse-opam  --build-only --ocaml-version 4.14.0 -l ./mirage.opam.locked
opam-monorepo: [ERROR] Cannot find any packages to vendor.
Either create some *.opam files in the local repository, or specify them manually via 'opam monorepo lock <packages>'.
make: *** [Makefile:67: lock] Error 1
```

I'll try your steps.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -8,6 +8,9 @@
 
 ### Fixed
 
+- Fix resolving refs of locally pinned repositories (#326, #332, @hannesm,
+  @Leonidas-from-XIV)
+
 ### Removed
 
 ### Security

--- a/lib/git.ml
+++ b/lib/git.ml
@@ -56,12 +56,19 @@ module Ls_remote = struct
           let result = search_ref (prefix ^ ref) parsed_lines in
           interpret_search_result result
         in
-        match (search "refs/tags/", search "refs/heads/") with
-        | Some _, Some _ -> Error `Multiple_such_refs
-        | Some commit, None | None, Some commit -> Ok commit
-        | None, None when is_commit ref parsed_lines -> Ok ref
-        | None, None when looks_like_commit ref -> log_approx ()
-        | None, None -> Error `No_such_ref)
+        match (search "", search "refs/tags/", search "refs/heads/") with
+        | Some _, Some _, Some _
+        | Some _, Some _, None
+        | Some _, None, Some _
+        | None, Some _, Some _ ->
+            Error `Multiple_such_refs
+        | Some commit, None, None
+        | None, Some commit, None
+        | None, None, Some commit ->
+            Ok commit
+        | None, None, None when is_commit ref parsed_lines -> Ok ref
+        | None, None, None when looks_like_commit ref -> log_approx ()
+        | None, None, None -> Error `No_such_ref)
 
   let parse_ref_output_line ~symref line =
     match String.extract_blank_separated_words line with

--- a/test/bin/pinning.t/opam-monorepo-test-other/opam-monorepo-test-other.opam
+++ b/test/bin/pinning.t/opam-monorepo-test-other/opam-monorepo-test-other.opam
@@ -1,0 +1,8 @@
+opam-version: "2.0"
+synopsis: "Dummy package for testing"
+maintainer: ["marek@tarides.com"]
+authors: ["Marek Kubica"]
+dev-repo: "git+http://example.com/opam-monorepo-test-other"
+depends: [
+  "dune"
+]

--- a/test/bin/pinning.t/run.t
+++ b/test/bin/pinning.t/run.t
@@ -1,0 +1,31 @@
+We have a simple project which depends on another package
+
+  $ opam show --just-file -fdepends ./this/this.opam
+  dune, opam-monorepo-test-other
+
+`git ls-remote` is a bit stupid and for some reason requires a valid `.git`
+folder and the dune cram tests don't have one. So let's create one here:
+
+  $ git init > /dev/null 2>&1
+
+Now let's create a git pin:
+
+  $ cd opam-monorepo-test-other
+  $ git init > /dev/null 2>&1
+  $ git add opam-monorepo-test-other.opam
+  $ git commit -m "Initial commit" > /dev/null
+  $ cd ..
+
+Let's add this pin and pin it to `HEAD`.
+
+  $ cd this
+  $ opam pin --yes --no-action add --kind git opam-monorepo-test-other '../opam-monorepo-test-other#HEAD' > /dev/null 2>&1
+
+Locking with this pin should work.
+
+  $ opam-monorepo lock > /dev/null 2>&1
+
+We need to clean up since this modifies global state.
+
+  $ opam pin remove opam-monorepo-test-other > /dev/null
+  $ cd ..

--- a/test/bin/pinning.t/this/this.opam
+++ b/test/bin/pinning.t/this/this.opam
@@ -1,0 +1,5 @@
+opam-version: "2.0"
+depends: [
+  "dune"
+  "opam-monorepo-test-other"
+]

